ecs-init/cache/dependencies.go

```diff
@@ -26,15 +26,103 @@ import (
 	"os"
 	"path/filepath"
 
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/s3"
 	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	log "github.com/cihub/seelog"
+	"github.com/pkg/errors"
 )
 
-// s3Downloader captures the only method used from the s3 client
-type s3Downloader interface {
+// s3API captures the only method used from the s3 package
+type s3API interface {
 	Download(w io.WriterAt, input *s3.GetObjectInput, options ...func(*s3manager.Downloader)) (n int64, err error)
 }
 
+type s3BucketDownloaderAPI interface {
+	download(fileName, cacheDir string, fs fileSystem) (string, error)
+}
+
+// s3BucketDownloader wraps a bucket together with a downloader that can download from it
+type s3BucketDownloader struct {
+	bucket string
+	region string
+	s3     s3API
+}
+
+func newS3BucketDownloader(region, bucketName string) (*s3BucketDownloader, error) {
+	session, err := session.NewSession(&aws.Config{
+		Credentials: credentials.AnonymousCredentials,
+		Region:      aws.String(region),
+	})
+	if err != nil {
+		log.Debugf("Got error when initializing downloader in region %s", region)
+		return nil, err
+	}
+
+	s3BucketDownloader := &s3BucketDownloader{
+		s3:     s3manager.NewDownloader(session),
+		bucket: bucketName,
+		region: region,
+	}
+
+	return s3BucketDownloader, nil
+}
+
+func (bd *s3BucketDownloader) download(fileName, cacheDir string, fs fileSystem) (string, error) {
+	file, err := fs.TempFile(cacheDir, fileName)
+	if err != nil {
+		return "", errors.Wrap(err, "could not create local file during download")
+	}
+
+	defer func() { // make sure we also handle possible error from f.Close()
+		cerr := file.Close()
+		if err == nil { // if no error return from download, captures error from f.Close()
+			err = cerr
+		}
+	}()
+
+	_, err = bd.s3.Download(file, &s3.GetObjectInput{
+		Bucket: aws.String(bd.bucket),
+		Key:    aws.String(fileName),
+	})
+
+	return file.Name(), err
+}
+
+type s3DownloaderAPI interface {
```

Comment:
hm do we really need to put this behind an interface?

Reply:
I think this one is needed so that i can write the test